compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPackUnPack.cpp

```diff
@@ -395,16 +393,110 @@ lowerOpWithEncoding(RewriterBase &rewriter, tensor::EmptyOp emptyOp,
       materializeEncodingInfo->innerDimsPos,
       materializeEncodingInfo->outerDimsPerm);
   Operation *newEmptyOp = rewriter.create<tensor::EmptyOp>(
-      loc, newShape, resultType.getElementType());
+      loc, newShape, emptyType.getElementType());
   return newEmptyOp;
 }
 
+/// Converts a linalg::GenericOp with encoded inputs into the packed domain.
+/// The `genericOp` must have all parallel iterator types and a single output
+/// with an identity indexing map.
+static FailureOr<Operation *>
+lowerGenericOpWithEncoding(RewriterBase &rewriter, linalg::GenericOp genericOp,
+                           ValueRange convertedInputOperands,
+                           ValueRange convertedOutputOperands,
+                           MaterializeEncodingFn materializeEncodingFn) {
+  if (!genericOp.hasPureTensorSemantics()) {
+    return failure();
+  }
+  if (genericOp.getNumReductionLoops() != 0) {
+    return rewriter.notifyMatchFailure(genericOp, "Loops are not all parallel");
+  }
+  if (genericOp.getNumDpsInits() != 1) {
+    return rewriter.notifyMatchFailure(genericOp, "Not only 1 init operand");
+  }
+  OpOperand *outputOperand = genericOp.getDpsInitOperand(0);
+  AffineMap outputMap = genericOp.getMatchingIndexingMap(outputOperand);
+  if (!outputMap.isIdentity()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "Output indexing map is not identity");
+  }
+  FailureOr<MaterializeEncodingInfo> outMaterializeEncodingInfo =
+      materializeEncodingFn(
+          cast<RankedTensorType>(outputOperand->get().getType()));
+  if (failed(outMaterializeEncodingInfo)) {
+    return rewriter.notifyMatchFailure(
+        genericOp, "MaterializeEncodingInfo failed for output");
+  }
+
+  auto convertedResultType =
+      cast<RankedTensorType>(convertedOutputOperands[0].getType());
+  SmallVector<utils::IteratorType> iteratorTypes(convertedResultType.getRank(),
+                                                 utils::IteratorType::parallel);
+  // Compute the new indexing maps for the packed layout. This assumes that
+  // the output map is identity, and that all iterator types are parallel.
```

Comment:
Why do we need the "iterator types are parallel" assumption? I thought the assumption is that all the indexing maps are projected permutation maps?

Reply:
I mostly made this assumption because we don't need to support reductions, and I didn't want to have to think about the reduction case. When we eventually want to support the reduction case, we can extend it.